File: src/main/java/io/github/jolkert/discordbotlink/jda/listener/LinkChannelListener.java
package io.github.jolkert.discordbotlink.jda.listener;

import io.github.jolkert.discordbotlink.DiscordBotLink_LEGACY;
import io.github.jolkert.discordbotlink.data.UserData;
import io.github.jolkert.discordbotlink.util.NameUtils;
import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.events.message.MessageReceivedEvent;
import net.dv8tion.jda.api.hooks.ListenerAdapter;
import net.minecraft.server.MinecraftServer;
import net.minecraft.text.*;

import java.util.Objects;

public class LinkChannelListener extends ListenerAdapter
{
	private static final String COMMAND_PREFIX = "./";

	@Override
	public void onMessageReceived(MessageReceivedEvent event)
	{
		if (event.getAuthor().isBot() || event.getAuthor().isSystem() || !event.getChannel().getId().equals(DiscordBotLink_LEGACY.Bot.getConfig().linkChannelId()))
			return;

		String message = event.getMessage().getContentRaw();
		MinecraftServer server = DiscordBotLink_LEGACY.Server;
		if (message.startsWith(COMMAND_PREFIX) && Objects.requireNonNull(event.getMember()).hasPermission(Permission.MANAGE_SERVER))
		{
			server.sendMessage(Text.of(NameUtils.nameWithDiscriminator(event.getAuthor()) + ": " + message));
			server.getCommandManager().executeWithPrefix(server.getCommandSource(), message.substring(COMMAND_PREFIX.length() - 1));
		}
		else
		{
			UserData author = DiscordBotLink_LEGACY.Bot.getUserData().getUser(event.getAuthor().getId());
			MutableText senderText = MutableText.of(new LiteralTextContent(NameUtils.getNickname(Objects.requireNonNull(event.getMember()))))
					.setStyle(Style.EMPTY
							.withColor(author.getRoleInfo().getTopColor())
							.withHoverEvent(HoverEvent.Action.SHOW_TEXT.buildHoverEvent(Text.of(author.getHoverText()))));

			MutableText finalMessage = MutableText.of(new LiteralTextContent("["))
					.append(SERVER_TEXT)
					.append("/")
					.append(senderText)
					.append("] " + message);

			server.getPlayerManager().broadcast(finalMessage, false);
		}

	}

	private static final MutableText SERVER_TEXT = MutableText.of(new LiteralTextContent("Server")).setStyle(Style.EMPTY.withColor(0xdabbed));
}

File: src/main/kotlin/io/github/jolkert/discordbotlink/jda/listener/LinkChannelListener.kt
package io.github.jolkert.discordbotlink.jda.listener

import io.github.jolkert.discordbotlink.DiscordBotLink
import io.github.jolkert.discordbotlink.DiscordBotLink.Bot
import io.github.jolkert.discordbotlink.DiscordBotLink.users
import io.github.jolkert.discordbotlink.extension.isPerson
import io.github.jolkert.discordbotlink.extension.nameWithDiscriminator
import io.github.jolkert.discordbotlink.extension.nicknameOrDefault
import io.github.jolkert.discordbotlink.jda.data.UserData
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import net.dv8tion.jda.api.hooks.ListenerAdapter
import net.minecraft.text.*
import java.util.*

object LinkChannelListener : ListenerAdapter()
{
	private const val MINECRAFT_COMMAND_PREFIX: String = "./"
	private val DUMMY_UUID = UUID.randomUUID()

	override fun onMessageReceived(event: MessageReceivedEvent)
	{
		if (event.channel.id != Bot.linkChannel?.id || !event.author.isPerson)
			return

		val authorAsMember = event.member
			?: return // smart cast wont work with `event.member` and i dont wanna put !! everywhere -morgan 2023-04-14
		val message = event.message.contentRaw
		val server = DiscordBotLink.Server

		if (message.startsWith(MINECRAFT_COMMAND_PREFIX) && authorAsMember.hasPermission(Permission.MANAGE_SERVER))
		{
			server.sendMessage(Text.of("${event.author.nameWithDiscriminator}: $message"))
			server.commandManager.executeWithPrefix(
				server.commandSource,
				message.substring(MINECRAFT_COMMAND_PREFIX.length - 1)
			)
		} else
		{
			val authorData = users[event.author.id] ?: UserData(
				minecraftData = UserData.MinecraftUserData("",	DUMMY_UUID), // dummy mc data cause we need it. maybe change that later -morgan 2023-04-14
				discordData = UserData.DiscordUserData(event.author),
				memberData = UserData.GuildMemberData.of(authorAsMember)
			)
			val senderText = MutableText.of(LiteralTextContent(authorAsMember.nicknameOrDefault))
				.setStyle(
					Style.EMPTY
						.withColor(authorData.memberData.topColor)
						.withHoverEvent(HoverEvent.Action.SHOW_TEXT.buildHoverEvent(Text.of(authorData.hoverText)))
				)

			val finalMessage = MutableText.of(LiteralTextContent("[")).append(senderText).append("] $message")

			server.playerManager.broadcast(finalMessage, false)
		}
	}
}